src/main/java/com/iota/iri/Snapshot.java

```diff
@@ -2,175 +2,177 @@
 import com.iota.iri.model.Hash;
-import org.apache.commons.io.IOUtils;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import java.io.*;
 import java.util.HashMap;
-import java.util.Iterator;
 import java.util.Map;
 import java.util.concurrent.locks.ReadWriteLock;
 import java.util.concurrent.locks.ReentrantReadWriteLock;
-import java.util.stream.Collectors;
 
 
-public class Snapshot {
+public final class Snapshot {
+
     private static final Logger log = LoggerFactory.getLogger(Snapshot.class);
+
     public static String SNAPSHOT_PUBKEY = "TTXJUGKTNPOOEXSTQVVACENJOQUROXYKDRCVK9LHUXILCLABLGJTIPNF9REWHOIMEUKWQLUOKD9CZUYAC";
     public static int SNAPSHOT_PUBKEY_DEPTH = 6;
     public static int SNAPSHOT_INDEX = 2;
     public static int SPENT_ADDRESSES_INDEX = 3;
     private static Snapshot initialSnapshot;
 
-
     public final ReadWriteLock rwlock = new ReentrantReadWriteLock();
+    private final Map<Hash, Long> state;
+    private int index;
 
 
+    /**
+     * {@code initialSnapshot} will only be assigned if this method succeeds. In the event of failure,
+     * the Java Virtual Machine will exit with -1.
+     *
+     * @param snapshotPath    the snapshot file
+     * @param snapshotSigPath the snapshot signature file
+     * @param testnet         true if using the testnet
+     * @return {@code initialSnapshot} if the Snapshot is read, verified, and assigned.
+     */
     public static Snapshot init(String snapshotPath, String snapshotSigPath, boolean testnet) {
-        //This is not thread-safe (and it is ok)
-        if (initialSnapshot == null) {
-            if (!testnet && !SignedFiles.isFileSignatureValid(snapshotPath, snapshotSigPath, SNAPSHOT_PUBKEY,
+        synchronized (Snapshot.class) {
+            if (initialSnapshot != null) {
+                return initialSnapshot;
+            }
+            try {
+                if (!testnet && !SignedFiles.isFileSignatureValid(snapshotPath, snapshotSigPath, SNAPSHOT_PUBKEY,
                     SNAPSHOT_PUBKEY_DEPTH, SNAPSHOT_INDEX)) {
-                throw new RuntimeException("Snapshot signature failed.");
+                    throw new IllegalStateException("Snapshot signature failed.");
+                }
+                Map<Hash, Long> initialState = initInitialState(snapshotPath);
+                checkStateHasCorrectSupply(initialState);
+                checkInitialSnapshotIsConsistent(initialState);
+                // Prune zero values - to honour original intent of 'remove'
+                // of ZERO values in original Snapshot.isConsistent(...)
+                initialState.entrySet().removeIf(entry -> entry.getValue() == 0);
+
+                return initialSnapshot = new Snapshot(initialState, 0);
+
+            } catch (Exception e) {
+                log.error("Quitting - Unable to verify snapshot: " + e, e);
+                System.exit(-1);
+                throw new Error("unreachable");
             }
```

Comment:
System.exit(-1) should really not be here .. or anywhere outside the IRI.main method. 
As part of initialization, it should be possible to throw an exception that can be captured at a strategic point so that the application can exit. This also limits testing because if a test was designed to make this fail, then the application could not properly test it since it exits.